### ISSUE TITLE
Failing to parse last message on network termination

### ISSUE THREAD

**Issue by dustin** [snoyberg/conduit#419]
This may be related to #278, but I've got a network client I've been developing with conduit which has a confusing behavior on server-initiated disconnect.

In a test client, I'm intentionally sending a bad protocol message which causes the server to disconnect me with a reason.  The servers sends a disconnect packet and then immediately resets the connection.  AFAICT, I never actually receive this final packet in my client.  I can see it on the wire, but it never touches my parser.

The code in question isn't that complicated (part of https://github.com/dustin/mqtt-hs):

```
      runConduit $ appSource ad
        .| conduitParser (parsePacket _protLvl)
        .| C.mapM_ (\(_,x) -> liftIO (dispatch c pch x))
```

This part of the client is bracketed and I can see that conduit bailed early, but the way this happens leads to a rather uninformative STM error in my client instead of a more useful error as communicated by the server.

Am I doing something wrong, or is it expected that an immediate reset after a packet may cause the packet to not be parsed?

**Comment by dustin** [comment]
<img width="960" alt="Screen Shot 2019-09-11 at 11 58 05" src="https://user-images.githubusercontent.com/1779/64726577-ce685f00-d48b-11e9-8983-e730bcf3eef5.png">

Screenshot of what I get from the server.  The RST is immediately after the last packet, so they're basically at the same time.  I'd still like those last bits of data if possible.

**Comment by snoyberg** [comment]
I don't believe the STM error is coming from conduit, I believe it's coming from other parts of your code. conduit would support reading the remaining data on the socket. Perhaps you have another thread which is trying to write to the now-closed socket. I'd recommend trying to reduce this to a minimal repro.
